sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -144,11 +153,16 @@ class Column(val expr: Expression) extends Logging {
   override def toString: String = toPrettySQL(expr)
 
   override def equals(that: Any): Boolean = that match {
-    case that: Column => that.expr.equals(this.expr)
+    case that: Column => that.normalizedExpr().equals(this.normalizedExpr())
     case _ => false
   }
 
-  override def hashCode: Int = this.expr.hashCode()
+  override def hashCode: Int = this.normalizedExpr().hashCode()
+
+  private def normalizedExpr(): Expression = expr match {
```

Comment:
When we add the metadata in addDataFrameIdToCol, we use transform. Do we need to use it here when removing it? 

Our test case can cover it?

Reply:
good catch!

However, I'm not able to test it. The `normalizedExpr` is only used in `equals` and `hashCode`. For `df("a.b")`, we will create a new alias everytime, see `AttributeSeq.resolve`.

That said, if we get multiple `Column` instances via calling `df("a.b")` multiple times, these `Column` instances will never equal to each other, because the alias is different.